common/djangoapps/third_party_auth/models.py

```diff
@@ -827,6 +848,77 @@ def get_setting(self, name):
         raise KeyError
 
+    def get_current_saml_configuration(self):
+        """
+        Get the current active SAMLConfiguration for this provider.
+        Falls back to the default configuration if none is set.
+
+        Provides observability through custom attributes regardless of toggle state.
+        """
+        # Always provide observability, regardless of toggle state
+        signal_handlers_enabled = ENABLE_SAML_CONFIG_SIGNAL_HANDLERS.is_enabled()
+        signal_handlers_status = 'enabled' if signal_handlers_enabled else 'disabled'
+        set_custom_attribute('saml_config.signal_handlers', signal_handlers_status)
+
+        # Check direct reference first
+        if self.saml_configuration:
+            # When signal handlers are enabled, trust the direct reference
+            if signal_handlers_enabled:
+                set_custom_attribute('saml_config.using', 'direct:id=' + str(self.saml_configuration.id))
+                return self.saml_configuration
+
+            # When signal handlers are disabled, check if we have a newer configuration
+            latest_config = self._get_latest_configuration()
+            if latest_config and latest_config.id != self.saml_configuration.id:
+                set_custom_attribute('saml_config.using', 'latest_found:id=' + str(latest_config.id))
+                set_custom_attribute('saml_config.outdated_reference', 'true:old_id=' + str(self.saml_configuration.id))
+                return latest_config
+            else:
+                # Either direct reference is still current, or latest lookup failed - use direct reference
+                if latest_config:
+                    set_custom_attribute('saml_config.using', 'direct:id=' + str(self.saml_configuration.id))
+                else:
+                    # Latest lookup failed, but we have a direct reference - use it with warning
+                    set_custom_attribute('saml_config.using', 'direct_fallback:id=' + str(self.saml_configuration.id))
+                    set_custom_attribute('saml_config.latest_lookup_failed', 'true')
+                return self.saml_configuration
+
+        # Fall back to default configuration
+        return self._get_default_configuration()
+
+    def _get_latest_configuration(self):
+        """Get the latest configuration for this provider's site and slug."""
+        # Protect against None saml_configuration
+        if not self.saml_configuration:
+            set_custom_attribute('saml_config.latest_lookup_error', 'no_saml_configuration')
+            return None
+
+        try:
+            return SAMLConfiguration.current(self.site_id, self.saml_configuration.slug)
+        except Exception as e:  # pylint: disable=broad-except
+            # Handle any database, configuration, or attribute errors
+            error_type = type(e).__name__
+            set_custom_attribute('saml_config.latest_lookup_error', error_type + ':' + str(e))
+            return None
```

Comment:
Based on preexisting code, I'm wondering if this code is safe and if we could greatly simplify.
```suggestion
        if not self.saml_configuration:
            return None

        return SAMLConfiguration.current(self.site_id, self.saml_configuration.slug)
```
I'd also leave `set_custom_attribute` off of these helper functions. It adds too much additional code and testing.

Reply:
Implemented